vowpalwabbit/cb_dro.cc

```diff
@@ -0,0 +1,155 @@
+#include "reductions.h"
+#include "cb_dro.h"
+#include "distributionally_robust.h"
+#include "explore.h"
+
+#include "rand48.h"
+
+using namespace LEARNER;
+using namespace VW;
+using namespace VW::config;
+
+namespace VW
+{
+struct cb_dro_data
+{
+  explicit cb_dro_data(double alpha, double tau, double wmax) : chisq(alpha, tau, 0, wmax) {}
+
+  template <bool is_learn, bool is_explore>
+  inline void learn_or_predict(multi_learner &base, multi_ex &examples)
+  {
+    // Some explanation required.
+    //
+    // VW currently is either
+    //    in prediction mode with exploration (aka --cb_explore_adf)
+    //    or in (off-policy) learning without exploration (aka --cb_adf)
+    //
+    // In practice, nobody seems to do
+    // "off policy learning of the stochastic exploration policy".
+    //
+    // Ergo, the following always optimizes the bound on the associated
+    // deterministic argmax score policy.
+
+    multiline_learn_or_predict<false>(base, examples, examples[0]->ft_offset);
+
+    if (is_learn)
+    {
+      const auto it = std::find_if(examples.begin(), examples.end(), [](example *item) { return !item->l.cb.costs.empty(); });
+
+      if (it != examples.end())
+      {
+        const CB::cb_class logged = (*it)->l.cb.costs[0];
+        const uint32_t labelled_action = std::distance(examples.begin(), it);
+
+        const auto action_scores = examples[0]->pred.a_s;
+
+        // cb_explore_adf => want maximum probability
+        // cb_adf => want minimum cost
```

Comment:
minimum cost is not necessarily defined.   It works when we have a CSOAA reduction, but there are others for which it does not.   Instead, can we just pick the preferred action of the cb_adf policy?

Reply:
now using action_scores.begin()